### PR TITLE
Rewrite local storage

### PR DESCRIPTION
**Pull request by akhenry**
Closes https://github.com/nasa/openmct/issues/4589
Closes https://github.com/nasa/openmct/issues/4590

* Re-implements the localStorage provider in vanilla JavaScript.

* Also removes the identifierService from all non-legacy usage.

* Some refactoring of the Couch object provider needed to acccommodate removal of the identifierService.
    * Object identifiers are no longer persisted in Couch at all.
    * Couch object provider is installed in both the 'mct' and '' namespaces. During development it was discovered that current deployments using the Couch adapter are producing objects with a mixture of namespaces (both '' and 'mct'), hence necessitating an object provider that handles both cases. This translation was previously done by the IdentifierService.

* This change may cause compatibility issues for any legacy custom persistence providers that were implicitly relying on the `identifierService` to perform translations between the '' and 'mct' namespaces. If this is the case, it is recommended that plugins are updated to use the current Open MCT API, which requires a single unambiguous namespace to be specified. 

* Any database-specific compatibility issues with existing data that may use a mixture of namespaces (as Couch did) should be handled in the plugin itself.  The Open MCT Couch DB plugin can be used as an example.

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots. - This change may cause issues with legacy persistence providers that relied on the IdentifierService to translate between the 'mct' and ''(default) namespaces.

### Author Checklist

* [X] Changes address original issue?
* [X] Unit tests included and/or updated with changes?
* [X] Command line build passes?
* [X] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### CODE REVIEW

File: src/plugins/persistence/couch/CouchObjectProvider.js [hunk 1]
```diff
@@ -557,11 +556,31 @@ class CouchObjectProvider {
     update(model) {
         let intermediateResponse = this.getIntermediateResponse();
         const key = model.identifier.key;
+        model = this.toPersistableModel(model);
```
Comment (shefalijoshi) [inline src/plugins/persistence/couch/CouchObjectProvider.js:559]: Were there issues where the model was being modified?

File: src/plugins/persistence/couch/plugin.js [hunk 2]
```diff
@@ -22,11 +22,15 @@
...
 
 export default function CouchPlugin(options) {
     return function install(openmct) {
         install.couchProvider = new CouchObjectProvider(openmct, options, NAMESPACE);
-        openmct.objects.addProvider(PERSISTENCE_SPACE, install.couchProvider);
+
+        // Unfortunately, for historical reasons, Couch DB produces objects with a mix of namepaces (alternately "mct", and "")
+        // Installing the same provider under both namespaces means that it can respond to object gets for both namespaces.
+        openmct.objects.addProvider(LEGACY_SPACE, install.couchProvider);
+        openmct.objects.addProvider(NAMESPACE, install.couchProvider);
```
Comment (shefalijoshi) [inline src/plugins/persistence/couch/plugin.js:34]: Interesting! Are we doing this to move away from the 'mct' namespace?

File: src/plugins/localStorage/plugin.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+import LocalStorageObjectProvider from './LocalStorageObjectProvider';
+
+export default function (namespace = '', storageSpace = 'mct') {
+    return function (openmct) {
+        openmct.objects.addProvider(namespace, new LocalStorageObjectProvider(storageSpace));
```
Comment (shefalijoshi) [inline src/plugins/localStorage/plugin.js:27]: Is the namespace '' because we're moving away from the 'mct' namespace in general?

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
This is cool!
I am concerned that getting rid of the default PERSISTENCE_SPACE will break object providers others have written. 
I believe VISTA also uses several namespaces for their objects, so we should make sure everything works for them after this change.

File: src/api/objects/ObjectAPI.js [hunk 4]
```diff
@@ -64,33 +61,17 @@ ObjectAPI.prototype.supersecretSetFallbackProvider = function (p) {
...
-    const keyString = utils.makeKeyString(identifier);
-    const identifierService = this.getIdentifierService();
-    const namespace = identifierService.parse(keyString).getSpace();
 
     if (identifier.key === 'ROOT') {
         return this.rootProvider;
     }
 
-    return this.providers[namespace] || this.fallbackProvider;
+    return this.providers[identifier.namespace] || this.fallbackProvider;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/api/objects/ObjectAPI.js:74]: By design, our API removed the distinction between Namespaces and Persistence spaces years ago when it was rewritten. The distinction creates a degree of unnecessary complexity that confuses even seasoned developers. 

The old way of defining object and persistence providers which supported this distinction has been deprecated for years, so I am fine with breaking backwards compatibility here.

File: src/plugins/localStorage/plugin.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*****************************************************************************
+ * Open MCT Web, Copyright (c) 2014-2015, United States Government
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/localStorage/plugin.js:2]: Good catch, thanks.

File: src/plugins/persistence/couch/CouchObjectProvider.js (hunk 1 above)
Comment (akhenry) [reply src/plugins/persistence/couch/CouchObjectProvider.js:559]: The only thing this line does is to strip identifiers from objects before they are persisted. Persisting our identifiers in Couch DB is redundant at best, and at worst could become stale. The `key` part of the identifier is just duplicating what we use to identify the document in Couch already, and the namespace is irrelevant (for the purposes of Couch DB - it's an Open MCT concept).

File: src/plugins/persistence/couch/plugin.js (hunk 2 above)
Comment (akhenry) [reply src/plugins/persistence/couch/plugin.js:34]: Yes, the mct namespace has no special meaning and translating between it and the default namespace just causes confusion. Other changes in this PR mean that the namespace is not persisted with Couch objects anyway, so in principle we could make this configurable in future.

File: src/plugins/localStorage/plugin.js (hunk 3 above)
Comment (akhenry) [reply src/plugins/localStorage/plugin.js:27]: Yes, the mct namespace never had any special meaning anyway. Some of our legacy providers mapped between the default namespace '' and the 'mct' namespace for reasons that never made a lot of sense to me and just addeded unneeded complexity.

**Comment by akhenry** [comment]
> I am concerned that getting rid of the default PERSISTENCE_SPACE will break object providers others have written.
It will break them, but supporting them means maintaining a translation between '' and 'mct' which was needlessly complex and confusing. In any case, legacy object providers have been deprecated for years, our current API has never supported the distinction between persistence and namespaces.

**Comment by akhenry** [comment]
> I believe VISTA also uses several namespaces for their objects
Agree, we will need to test this. To be clear, this change does not prevent the use of multiple different namespaces, that's essential to maintain. It just removes the automatic translation between the 'mct' and '' namespaces which only ever existed to support legacy providers which were deprecated years ago.

File: src/plugins/localStorage/plugin.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/localStorage/plugin.js]: Fixed

**Review by shefalijoshi — APPROVED** [review]
Looks good
### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Commit messages meet standards?
* [ ] Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled `bug`? (only applicable if this PR is for a bug fix)
* [ ] List of Acceptance Tests Performed.

**Review by shefalijoshi — APPROVED** [review]

**Pull request merged**